src/kopls.js

```diff
@@ -0,0 +1,248 @@
+
+            result = new SingularValueDecomposition(TpiPrime.mmul(Matrix.sub(KX[i][i], Tp[i].mmul(TpiPrime))).mmul(Tp[i]), {
+                computeLeftSingularVectors: true,
+                computeRightSingularVectors: false
+            });
+            var CoTemp = result.leftSingularVectors;
+            var SoTemp = result.diagonalMatrix;
+
+            co[i] = CoTemp.subMatrix(0, CoTemp.rows - 1, 0, 0);
+            so[i] = SoTemp[0][0];
+
+            to[i] = Matrix.sub(KX[i][i], Tp[i].mmul(TpiPrime)).mmul(Tp[i]).mmul(co[i]).mul(Math.pow(so[i], -0.5));
+
+            var toiPrime = to[i].transpose();
+            toNorm[i] = Matrix.sqrt(toiPrime.mmul(to[i]));
+
+            to[i] = to[i].divRowVector(toNorm[i]);
+
+            var ITo = Matrix.sub(I, to[i].mmul(to[i].transpose()));
+
+            KX[0][i + 1] = KX[0][i].mmul(ITo);
+            KX[i + 1][i + 1] = ITo.mmul(KX[i][i]).mmul(ITo);
+        }
+
+        var lastTp = Tp[this.orthogonalComp] = KX[0][this.orthogonalComp].transpose().mmul(Up).mmul(SpPow);
+
+        var lastTpPrime = lastTp.transpose();
+        Bt[this.orthogonalComp] = inverse(lastTpPrime.mmul(lastTp)).mmul(lastTpPrime).mmul(Up);
+
+        this.Cp = Cp;
+        this.SpPow = SpPow;
+        this.Up = Up;
+        this.Tp = Tp;
+        this.co = co;
+        this.so = so;
+        this.to = to;
+        this.toNorm = toNorm;
+        this.Bt = Bt;
+        this.K = KX;
+    }
+
+    /**
+     * Predicts the output given the matrix to predict.
+     * @param {Matrix|Array} toPredict
+     * @return {Matrix} predictions
+     */
+    predict(toPredict) {
+
+        var KteTr = this.kernel.compute(toPredict, this.X);
+
+        var KteTrMc = KteTr;
+        KteTr = new Matrix(this.orthogonalComp + 1, this.orthogonalComp + 1);
+        KteTr[0][0] = KteTrMc;
+
+        var to = new Array(this.orthogonalComp);
+        var Tp = new Array(this.orthogonalComp);
+
+        var i;
+        for (i = 0; i < this.orthogonalComp; ++i) {
+            Tp[i] = KteTr[i][0].mmul(this.Up).mmul(this.SpPow);
+
+            to[i] = Matrix.sub(KteTr[i][i], Tp[i].mmul(this.Tp[i].transpose())).mmul(this.Tp[i]).mmul(this.co[i]).mul(Math.pow(this.so[i], -0.5));
+
+            to[i] = to[i].divRowVector(this.toNorm[i]);
+
+            var toiPrime = this.to[i].transpose();
+            KteTr[i + 1][0] = Matrix.sub(KteTr[i][0], to[i].mmul(toiPrime).mmul(this.K[0][i].transpose()));
+
+            var p1 = Matrix.sub(KteTr[i][0], KteTr[i][i].mmul(this.to[i]).mmul(toiPrime));
+            var p2 = to[i].mmul(toiPrime).mmul(this.K[i][i]);
+            var p3 = p2.mmul(this.to[i]).mmul(toiPrime);
+
+            KteTr[i + 1][i + 1] = p1.sub(p2).add(p3);
+        }
+
+        Tp[i] = KteTr[i][0].mmul(this.Up).mmul(this.SpPow);
+        var prediction = Tp[i].mmul(this.Bt[i]).mmul(this.Cp.transpose());
+
+        this.predScoreMat = Tp;
+        this.predYOrthVectors = to;
```

Comment:
I don't really like this approach. Making a prediction shouldn't change the state of the predictor

Reply:
so, should I return those values inside of the return statement?